src/caffe/layers/cosine_loss_layer.cpp

```diff
@@ -0,0 +1,90 @@
+#include <algorithm>
+#include <cmath>
+#include <vector>
+
+#include "caffe/layers/cosine_loss_layer.hpp"
+#include "caffe/util/math_functions.hpp"
+
+namespace caffe {
+
+template <typename Dtype>
+void CosineLossLayer<Dtype>::Reshape(
+    const vector<Blob<Dtype>*>& bottom, const vector<Blob<Dtype>*>& top) {
+  LossLayer<Dtype>::Reshape(bottom, top);
+  CHECK_EQ(bottom[0]->count(), bottom[1]->count())
+      << "Inputs must have the same dimension.";
+  vector<int> shape_vec;
+  shape_vec.push_back(bottom[0]->num());
+  dots_.Reshape(shape_vec);
+  len_inp_.Reshape(shape_vec);
+  len_label_.Reshape(shape_vec);
+}
+
+template <typename Dtype>
+void CosineLossLayer<Dtype>::Forward_cpu(
+    const vector<Blob<Dtype>*>& bottom, const vector<Blob<Dtype>*>& top) {
+  const Dtype* inp_data = bottom[0]->cpu_data();
+  const Dtype* label_data = bottom[1]->cpu_data();
+  Dtype* len_inp_data = len_inp_.mutable_cpu_data();
+  Dtype* len_label_data = len_label_.mutable_cpu_data();
+  Dtype* dots_data = dots_.mutable_cpu_data();
+  int num = bottom[0]->num();
+  int dim = bottom[0]->count()/bottom[0]->num();
+
+  // compute lengths and dot products for the batch
+  for (int i = 0; i < bottom[0]->num(); ++i) {
+    len_inp_data[i] = caffe_cpu_nrm2(dim, inp_data + (i*dim));
+    len_label_data[i] = caffe_cpu_nrm2(dim, label_data + (i*dim));
+    dots_data[i] = caffe_cpu_dot(dim, inp_data + (i*dim),
+                                 label_data + (i*dim));
+  }
+
+  Dtype loss(0.0);
+  for (int i = 0; i < num; ++i) {
+    loss += 1 - dots_data[i] / (len_inp_data[i]*len_label_data[i]);
+  }
+  top[0]->mutable_cpu_data()[0] = loss / num;
+}
+
+template <typename Dtype>
+void CosineLossLayer<Dtype>::Backward_cpu(
+    const vector<Blob<Dtype>*>& top, const vector<bool>& propagate_down,
+    const vector<Blob<Dtype>*>& bottom) {
+  if (propagate_down[1]) {
+    LOG(FATAL) << this->type()
```

Comment:
Indeed, the `CosineEmbeddingLoss` seems more robust. Nevertheless, since this layer is symmetric, you should allow gradients to be backprop to both inputs. See, e.g., [`"EuclideanLoss"`][1] layer: the layer is symmetric w.r.t its inputs and thus it allows for backprop to **both** `bottom`s according to `propagate_down` indicator.

 [1]: https://github.com/BVLC/caffe/blob/master/src/caffe/layers/euclidean_loss_layer.cpp#L35